src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
+
+type 'a t = {
+  mutable head : 'a node ref;
+  mutable tail : 'a node ref;
```

Comment:
The `head` and `tail` fields are mutated by `pop` and `push` operations, respectively.  Now that both fields are next to each other, it means there is very high likelyhood of [false sharing](https://en.wikipedia.org/wiki/False_sharing), which will degrade performance.